src/index.js

```diff
@@ -0,0 +1,360 @@
+
+  /**
+   * Create a new MongoOplogReader
+   * @param options
+   *  - connectionStrings {String[]} a list of mongodb connection strings
+   *  - redisClient {Object} an instance of the 'redis' module
+   *  - redundancy {Integer} number of processes per shard
+   *  - ttl {Integer} max time it may take a cluster to recover (in seconds)
+   *  - keyPrefix {String} the redis key prefix (default: 'mongoOplogReader')
+   */
+  constructor(options) {
+    super();
+    if (!options.redisClient || !options.redisClient.sadd) {
+      throw new Error('Invalid redisClient.');
+    }
+    this.connectionStrings = options.connectionStrings;
+    this.redisClient = Promise.promisifyAll(options.redisClient);
+    this.lock = RedisLock.createLock(options.redisClient);
+    this.oplogs = {};
+    this.replSets = {};
+    this.redundancy = options.redundancy || defaults.redundancy;
+    this.keyPrefix = options.keyPrefix || defaults.keyPrefix;
+    this.ttl = options.ttl || defaults.ttl;
+    this.masterDuration = options.masterDuration || defaults.masterDuration;
+    this.healthcheckDuration = options.healthcheckDuration || defaults.healthcheckDuration;
+    this.mongoConnectOptions = options.mongoConnectOptions || defaultMongoConnectOptions;
+    this.workerId = `${Math.random()}`.substr(2);
+    this.workerRegistrationKey = `${this.keyPrefix}:workerIds`;
+    this.assignmentsByConnStr = {};
+    this.assignmentsByWorkerId = {};
+
+    if (!Number.isInteger(this.redundancy) || this.redundancy > 10) {
+      throw new Error(`Specified redundancy '${this.redundancy}' should be an integer less than 10.`);
+    }
+  }
+
+  start() {
+    if (!Array.isArray(this.connectionStrings) || !this.connectionStrings.length) {
+      throw new Error('There are no connectionStrings.');
+    }
+    return Promise.resolve()
+      .then(() => this.registerWorker())
+      .then(() => this.acquireMasterLock())
+      .then(() => this.startPolling());
+  }
+
+  startPolling() {
+    // try to acquire the master lock every 15-30 seconds by default
+    const masterMs = this.masterDuration * 500;
+    this.masterInterval = setInterval(() => {
+      const randomDelay = Math.floor(masterMs * Math.random());
+      setTimeout(() => this.acquireMasterLock(), randomDelay);
+    }, masterMs);
+
+    // re-register this worker every 10 seconds by default
+    const healthcheckMs = this.healthcheckDuration * 1000;
+    this.healthcheckInterval = setInterval(() => this.registerWorker(), healthcheckMs);
+  }
+
+  /**
+   * If the "master" lock is acquired, this worker will be responsible for
+   * assigning oplogs to workers
+   */
+  acquireMasterLock() {
+    const key = `${this.keyPrefix}:master`;
+    return this.lock.acquire(key, {
+      timeout: this.masterDuration * 1000,
+      retries: 0
+    })
+      .then(() => {
+        debug(`${this.workerId} is master.`);
+        return this.assignWorkers();
+      })
+      // ignore lock acquisition errors
+      .catch(err => {});
```

Comment:
Even if we are not doing anything with it, I think we should log it.

Reply:
All workers compete for acquiring the master lock, so this will be a continuous stream of errors until the 30s TTL is reached and a new worker acquires the lock.